### DISCUSSION TITLE
How to impersonate a user?

### DISCUSSION THREAD

**Discussion by j-manu** [janko/rodauth-rails#87] [Q&A] [answered]
Sometimes an Admin User need to log in as regular user to troubleshoot. In devise I can use [bypass_sign_in](https://www.rubydoc.info/github/plataformatec/devise/Devise%2FControllers%2FSignInOut:bypass_sign_in) to achieve this. How can I do it in rodauth-rails assuming there are two configurations in RoadAuth - admin & account

**Comment by janko** [comment, marked as answer]
I suppose the easiest way would be to set the `:account_id` manually in the session, which is what Rodauth does when it logs the user in, and you can also set `:authenticated_by` to indicate impersonation (that's primarily for introspection purposes):

```rb
session[:account_id] = account.id
session[:authenticated_by] = [:admin]
```

This assumes the "account" configuration uses the default session key names, i.e. doesn't use a session key prefix. Note that I haven't tested this, so maybe there is something I overlooked.

The rodauth object also has a `#login` method, but that one resets the session on login, which from my understanding would log the admin out. So the previous approach is probably safer.

**Reply by j-manu** [reply]
This works!

Only wrinkle is that logging out the impersonated user logs out the admin too but I can live with that

**Comment by mikereczek** [comment]
@j-manu - Would [this](https://github.com/adam12/rodauth-become_account) be helpful?

I've never used it, but I remembered seeing it on the main rodauth documentation page under "External Features"

**Comment by j-manu** [comment]
@reczy That works if you are a user and switching to a user - https://github.com/adam12/rodauth-become_account/blob/master/test/test_become_account.rb#L105 - I'm not sure if it works when there are different configurations - Admin & User

@janko Thank you. I will test that.
